### PR TITLE
feat(observability): listening session funnel telemetry, cost guardrails, and debug queries

### LINKED ISSUE #154: [Listening Sessions] Add observability, fallback telemetry, and cost guardrails
## Parent
- Part of #105

## Problem
Long-session audio + multi-stage LLM processing can fail silently or become cost-prohibitive without hard observability.

## Scope
- Instrument end-to-end funnel events for listening sessions.
- Add provider metrics: latency, error class, fallback usage, transcript length, per-session cost estimate.
- Add alerting thresholds for spike in failures and synthesis timeout rates.
- Add admin/debug query path for failed sessions with stage breakdown.
- Add budget guardrails (daily/session caps + graceful degradation mode).

## Acceptance Criteria
- [ ] Dashboard/query path can answer: where sessions fail, how often, and at what cost.
- [ ] Alerts configured for critical failure spikes.
- [ ] Fallback and degraded-mode activations are visible and auditable.
- [ ] No sensitive transcript content is logged in plaintext.

## Out of Scope
- Growth analytics unrelated to listening sessions.

### PR DESCRIPTION
**Pull request by phrazzld**
## Summary

Closes #154

Long-session audio processing was a black box: silent failures gave no signal on *where* in the pipeline things broke or *how much* they cost. This PR instruments the full transcribe → synthesize funnel with provider metrics, per-session cost estimates, structured cost guardrails, and an admin debug query path — making it possible to answer "where do sessions fail, how often, and at what cost?"

## Changes

**Schema** (`convex/schema.ts`)
- Added 7 new optional fields to `listeningSessions`: `transcribeLatencyMs`, `synthesisLatencyMs`, `transcribeFallbackUsed`, `synthesisProvider`, `degradedMode`, `estimatedCostUsd`, `failedStage`

**Convex backend** (`convex/listeningSessions.ts`)
- `failListeningSession`: accepts `failedStage` to record which pipeline stage broke
- `completeListeningSession`: accepts `estimatedCostUsd` and clamps to ≥0
- `getDebugStats` (internal query): aggregates failure stage breakdown, fallback usage, degraded mode activations, total cost across last 100 failed sessions
- `getDailyUserCost` (internal query): per-user daily cost rollup for budget cap enforcement

**Synthesis API** (`app/api/listening-sessions/synthesize/route.ts`)
- `estimateCostUsd()`: token-based cost estimation from model family rate table
- `getUsageTokens()`: normalizes OpenAI/Anthropic usage response shapes
- `logSessionCostGuardrails()`: emits structured JSON error/warn logs when session cost exceeds `SESSION_COST_HARD_CAP_USD` ($0.50) or `SESSION_COST_WARN_USD` ($0.10)
- Returns `estimatedCostUsd` in response payload; passes `sessionId` through to Convex on complete

**Transcribe API** (`app/api/listening-sessions/transcribe/route.ts`)
- Records `transcribeLatencyMs` and `transcribeFallbackUsed` (ElevenLabs → Deepgram) via best-effort telemetry mutation
- Tags failures with `failedStage: "transcribing"` on error path

**Recorder hook** (`components/notes/useListeningSessionRecorder.ts`)
- Threads `sessionId` into synthesis request; propagates `estimatedCostUsd` back to Convex on complete
- Tracks `failedStage` variable across the pipeline for accurate error attribution

**Alert thresholds** (`lib/listening-sessions/alert-thresholds.ts`)
- Single constants file for session cost thresholds used by guardrails

**Guardrail tests** (`scripts/session-guardrails.ts`)
- 3 new checks: `checkObservabilityFields`, `checkFailedStageRecorded`, `checkCostEstimateStored`

**CLI** (`scripts/obs`)
- `obs session-stats [--limit N]`: queries Sentry for unresolved listening-session errors; renders count table + stage breakdown

## Acceptance Criteria

- [x] Dashboard/query path can answer: where sessions fail, how often, and at what cost (`getDebugStats`, `obs session-stats`)
- [x] Alerts configured for critical failure spikes (structured error logs at cost thresholds; `ALERT_THRESHOLDS` constants ready for Sentry alert rule binding)
- [x] Fallback and degraded-mode activations are visible and auditable (`transcribeFallbackUsed`, `degradedMode` persisted per session)
- [x] No sensitive transcript content is logged in plaintext (telemetry logs only `userIdSuffix`, `sessionId`, latency, provider, cost — never transcript text)

## Manual QA

**Cost guardrails (local dev):**
```bash
bun run dev
# Trigger a listening session with a long transcript via the UI
# In Vercel/terminal logs, look for:
# {"msg":"listening_session_cost_elevated",...} at warn level
# {"msg":"listening_session_cost_cap_exceeded",...} at error level (>$0.50)
```

**Debug stats query (Convex dashboard):**
```
# Go to Convex dashboard → Functions → listeningSessions.getDebugStats
# Run with no args; verify stageBreakdown, totalFailed, totalEstimatedCostUsd fields present
```

**Session-stats CLI:**
```bash
export SENTRY_AUTH_TOKEN=<your-token>
./scripts/obs session-stats --limit 10
# Expected: table of errors by count + stage breakdown (transcribing / synthesizing / upload)
# If no errors: "✓ No unresolved listening-session errors"
```

**Guardrail test suite:**
```bash
CONVEX_URL=... CONVEX_DEPLOY_KEY=... bun run scripts/session-guardrails.ts
# All checks should pass including the 3 new ones:
# ✓ Observability fields persistence
# ✓ Failed stage recorded
# ✓ Cost estimate stored
```

## Test Coverage

- `__tests__/listening-sessions/cost-estimation.test.ts` — 17 tests (estimateCostUsd + getUsageTokens)
- `__tests__/listening-sessions/cost-guardrails.test.ts` — 9 tests (threshold boundaries, edge cases)
- `__tests__/listening-sessions/fallback-artifacts.test.ts` — 13 tests (sentence extraction, quotes, context-aware hints)
- `__tests__/listening-sessions/pipeline-stages.test.ts` — 2 tests (stage order, type safety)
- `scripts/session-guardrails.ts` — 6 integration checks (to be migrated to vitest per #197)

## Before / After (Polish Pass)

**Before:** Working implementation with duplicated ~80-line synthesis schema in two files, untyped pipeline stage magic strings across 5 call sites, 4 unused alert threshold constants, untested production fallback path, and no clamping on cost estimation inputs.

**After:**
- **Extracted shared modules:** `synthesisSchema.ts` (prevents silent divergence between API route and background worker), `fallback-artifacts.ts` (testable fallback), `pipeline-stages.ts` (typed stage enum)
- **Typed pipeline stages:** `PipelineStage` union replaces magic strings; `satisfies` enforces at compile time
- **Cost estimation clamped:** `Math.max(0, cost)` prevents negative cost from malformed token inputs
- **Unused thresholds cleaned:** 4 dead constants removed, 2 used ones retained with documented future plans
- **21 new edge-case tests:** negative tokens, conflicting provider fields, fallback artifact extraction, cost=0/negative boundaries, pipeline stage ordering
- **3 arch issues created:** #195 (markSynthesizing overload), #196 (dead debug queries), #197 (migrate guardrails to vitest)
- **Quality gates:** typecheck ✓, lint ✓, 565/565 tests pass ✓

### CODE REVIEW

File: app/api/listening-sessions/synthesize/route.ts [hunk 1]
```diff
@@ -140,6 +142,76 @@ function makeFallbackArtifacts(transcript: string, context?: SynthesisContext):
...
+      : typeof totalTokensRaw === "number" && Number.isFinite(totalTokensRaw)
+        ? Math.max(0, Math.floor(totalTokensRaw))
+        : 0;
+
+  const completionTokens =
+    typeof completionTokensRaw === "number" && Number.isFinite(completionTokensRaw)
+      ? Math.max(0, Math.floor(completionTokensRaw))
+      : 0;
+
+  return { promptTokens, completionTokens };
```
[the remark this answers is not in the document]
Comment (phrazzld) [reply app/api/listening-sessions/synthesize/route.ts]: Fixed in bc7db5f. Reordered to compute `completionTokens` first, then derive `promptTokens` as `max(floor(totalTokens) - completionTokens, 0)` when the prompt-specific field is absent — eliminating the double-count.

File: convex/listeningSessions.ts [hunk 2]
```diff
@@ -428,6 +442,73 @@ export const getForProcessing = internalQuery({
...
+    const sessions = await ctx.db
+      .query("listeningSessions")
+      .withIndex("by_user", (q) => q.eq("userId", userId))
+      .collect();
+    const inRange = sessions.filter(
+      (session) =>
+        session.updatedAt >= dayStartMs &&
+        session.updatedAt <= dayEndMs &&
+        session.estimatedCostUsd !== undefined,
+    );
```
[the remark this answers is not in the document]
Comment (phrazzld) [reply convex/listeningSessions.ts]: Fixed in bc7db5f. Added `by_user_updatedAt` index `(["userId", "updatedAt"])` to the schema, then rewrote `getDailyUserCost` to use `.withIndex("by_user_updatedAt", q => q.eq("userId", userId).gte("updatedAt", dayStartMs))` with a DB-side `.filter()` for the upper bound. Eliminates the full-scan + JS array filter.

File: app/api/listening-sessions/synthesize/route.ts [hunk 3]
```diff
@@ -140,6 +142,76 @@ function makeFallbackArtifacts(transcript: string, context?: SynthesisContext):
...
+
+  const promptTokensRaw = asRecord.prompt_tokens ?? asRecord.input_tokens;
+  const completionTokensRaw = asRecord.completion_tokens ?? asRecord.output_tokens;
+  const totalTokensRaw = asRecord.total_tokens;
+
+  const promptTokens =
+    typeof promptTokensRaw === "number" && Number.isFinite(promptTokensRaw)
+      ? Math.max(0, Math.floor(promptTokensRaw))
+      : typeof totalTokensRaw === "number" && Number.isFinite(totalTokensRaw)
+        ? Math.max(0, Math.floor(totalTokensRaw))
```
[the remark this answers is not in the document]
Comment (phrazzld) [reply app/api/listening-sessions/synthesize/route.ts]: Fixed in bc7db5f (same fix as the gemini thread). `completionTokens` computed first; `promptTokens` fallback is `max(floor(totalTokens) - completionTokens, 0)`.

File: convex/listeningSessions.ts [hunk 4]
```diff
@@ -428,6 +442,73 @@ export const getForProcessing = internalQuery({
...
+      })),
+    };
+  },
+});
+
+export const getDailyUserCost = internalQuery({
+  args: { userId: v.id("users"), dayStartMs: v.number(), dayEndMs: v.number() },
+  handler: async (ctx, { userId, dayStartMs, dayEndMs }) => {
+    const sessions = await ctx.db
+      .query("listeningSessions")
```
[the remark this answers is not in the document]
Comment (phrazzld) [reply convex/listeningSessions.ts]: Fixed in bc7db5f. Added `by_user_updatedAt` index (`["userId", "updatedAt"]`) to schema and rewrote `getDailyUserCost` to use index range + DB-side `.filter()`. No more full-scan collect.

File: app/api/listening-sessions/synthesize/route.ts [hunk 5]
```diff
@@ -140,6 +142,76 @@ function makeFallbackArtifacts(transcript: string, context?: SynthesisContext):
   });
 }
 
+function estimateCostUsd(model: string, promptTokens: number, completionTokens: number): number {
+  const COST_PER_1M: Record<string, { input: number; output: number }> = {
```
[the remark this answers is not in the document]
Comment (phrazzld) [reply app/api/listening-sessions/synthesize/route.ts]: Fixed in 25bf266. Same fix — `estimateCostUsd` extracted and covered with tests for each model family, the unknown-model fallback, zero tokens, and a realistic small-session sanity check.

File: app/api/listening-sessions/synthesize/route.ts [hunk 6]
```diff
@@ -140,6 +142,76 @@ function makeFallbackArtifacts(transcript: string, context?: SynthesisContext):
...
+        msg: "listening_session_cost_elevated",
+        sessionId,
+        estimatedCostUsd: params.estimatedCostUsd,
+        model: params.model,
+        warnThresholdUsd: ALERT_THRESHOLDS.SESSION_COST_WARN_USD,
+        level: "warn",
+        timestamp: new Date().toISOString(),
+      }),
+    );
+  }
```
[the remark this answers is not in the document]
Comment (phrazzld) [reply app/api/listening-sessions/synthesize/route.ts]: Fixed in 25bf266. `logSessionCostGuardrails` now uses `log("error", ...)` and `log("warn", ...)` throughout — consistent with the rest of the route, JSON serialization and timestamps handled by the observability wrapper.

File: convex/listeningSessions.ts [hunk 7]
```diff
@@ -428,6 +442,73 @@ export const getForProcessing = internalQuery({
...
+        failedStage: session.failedStage ?? "unknown",
+        lastError: session.lastError,
+        retryCount: session.retryCount ?? 0,
+        transcribeFallbackUsed: session.transcribeFallbackUsed ?? false,
+        degradedMode: session.degradedMode ?? false,
+        estimatedCostUsd: session.estimatedCostUsd ?? 0,
+        updatedAt: session.updatedAt,
+      })),
+    };
+  },
```
[the remark this answers is not in the document]
Comment (phrazzld) [reply convex/listeningSessions.ts]: Fixed in 25bf266. `getDebugStats` now requires a `userId` arg and uses the `by_user_status` index to scope all queries to that user. Removed `userId` and `bookId` from `recentFailures` — returns only operational diagnostics (stage, error, retries, cost, timestamps).

File: lib/listening-sessions/alertThresholds.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+ * - COST_SPIKE_USD: Alert when per-user cost exceeds this in 1 day
+ */
+export const ALERT_THRESHOLDS = {
+  HIGH_FAILURE_RATE_PERCENT: 20, // 20% failure rate = critical alert
+  SYNTHESIS_TIMEOUT_RATE_PERCENT: 15, // 15% synthesis timeout rate = alert
+  SESSION_COST_WARN_USD: 0.1, // Per-session warning threshold
+  SESSION_COST_HARD_CAP_USD: 0.5, // Per-session hard cap (log error)
+  DAILY_USER_COST_WARN_USD: 1.0, // Per-user daily cost warning
+  DAILY_USER_COST_CAP_USD: 5.0, // Per-user daily cost cap
+} as const;
```
[the remark this answers is not in the document]
Comment (phrazzld) [reply lib/listening-sessions/alertThresholds.ts]: Fixed in 25bf266. Renamed to `alert-thresholds.ts`, updated the import in `synthesize/route.ts`, and replaced the stale `COST_SPIKE_USD` doc line with the actual exported keys.

File: convex/listeningSessions.ts [hunk 9]
```diff
@@ -506,46 +598,103 @@ export const markTranscribingHandler = async (
...
+  }
+  if (args.transcribeFallbackUsed !== undefined) {
+    patch.transcribeFallbackUsed = args.transcribeFallbackUsed;
+  }
   await ctx.db.patch(session._id, patch);
+  return shouldScheduleProcessing;
 };
 
 export const markTranscribing = mutation({
+  args: markTranscribingArgs,
```
[the remark this answers is not in the document]
Comment (phrazzld) [reply convex/listeningSessions.ts]: Fixed in b4bf676. The background worker now imports `estimateCostUsd`/`getUsageTokens` from the new `cost-estimation` module, times the synthesis call (`synthesisLatencyMs`), extracts token counts + resolved model from the raw OpenRouter response, and passes `estimatedCostUsd` to `completeInternal`. Structured log emitted on success with latency, model, tokens, and cost.

File: app/api/listening-sessions/synthesize/route.ts [hunk 10]
```diff
@@ -272,6 +350,22 @@ export const POST = withObservability(async (request: Request) => {
...
     const resolvedModel = raw.model ?? config.model;
+    const { promptTokens, completionTokens } = getUsageTokens(raw.usage);
+    const estimatedCostUsd = estimateCostUsd(resolvedModel, promptTokens, completionTokens);
+    const synthesisLatencyMs = Date.now() - synthesisStart;
+
+    logSessionCostGuardrails({
+      sessionId: body.sessionId,
+      estimatedCostUsd,
+      model: config.model,
+    });
```
[the remark this answers is not in the document]
Comment (phrazzld) [reply app/api/listening-sessions/synthesize/route.ts]: Fixed in 8f90f25001eedc9818fba608f1b23cb93fb20f96. `logSessionCostGuardrails` now receives `model: resolvedModel` (the actual model returned by OpenRouter) instead of `model: config.model` (the requested model).
